### ISSUE TITLE
Standard library support

### ISSUE THREAD

**Issue by zillemarco** [mono/CppSharp#749]
I know that the C++ standard library is supported only for C++/CLI and it's still experimental but I'd like to know how far are we from supporting it also on C#.
I'd like to know this becase I'd like to create C# bindings for one of my libraries which heavily uses the standard library (mostly vectors, strings and maps).

**Comment by ddobrev** [comment]
Hello @zillemarco . We are aware of the problem but the basic STD types are really difficult to support because they are templates (std::string is actually a type definition of basic_string<charT, traits, Alloc>). I have worked in that are but there has been little result so far. Up until a week I thought we at least has proper support for std::string but it turned our I was relying on binary symbols that are not always available on all platforms - among other problems. I am now busy with inlines because the approach we have used so far is unreliable. As soon as I am done which I hope is by the middle of next month, I'll fix our support for std::string which is a special case in our generator until the complete support for templates, which I have planned after that, is finished.

**Comment by zillemarco** [comment]
Yeah, I understand this will be a tough one. I didn't mean to hurry you since I don't need this right now but will be very helpful to me later on :) I just wanted to know how far you are from supporting STD :)

**Comment by MikeWMiller** [comment]
Hello ddobrev. Are you still planning on supporting the std library? Is there a way to tell the parser to skip it?  I had found driver.AddTranslationUnitPass(new IgnoreStdFieldsPass()); but it doesn't seem to exist anymore. Also do you have marshaling code for std::string? 

Thank you in advance, I am trying to figure out a way to use this for a large code base that we are wrapping C++ abstract classes around. Then we plan on wrapping those classes to expose them to c# via C++/CLI

**Comment by tritao** [comment]
@MikeWMiller We support `std::vector`, `std::map` and `std::string` on C++/CLI.

About `IgnoreStdFieldsPass` I think we got rid of it some time ago, we didn't realize that someone was using it. We could bring it back, feel free to open an issue requesting it.

You can find string marshaling code here: https://github.com/mono/CppSharp/blob/master/src/Generator/Types/Std/Stdlib.cs#L272

**Comment by MikeWMiller** [comment]
I have a method defined that takes a const std::wstring& and with this parameter in I get a generated file but the method has been stripped out. Do I need to do something to enable the STL strings? If I remove the parameter the method gets generated.

Also, it doesn't look like std::map is supported in c++/cli as the mthods for marshaling CLI in public class Map : TypeMap are not implemented (CLIMarshalToNative() and CLIMarshalToManaged()) - the ones for vector are filled out. am I missing something? I don't think std::map and System::Collections::Generic::Dictionary are blitable are they?

Oh I don't need the IgnoreStdFieldsPass - I didn't have the header file included before so it was giving me a parsing error.

I think this project has great promise, it is so customizable it's hard to figure out all the options, I just need to keep playing with it.

Thank you for your previous and prompt reply.

**Comment by tritao** [comment]
Check this PR for `std::wstring`, it could use some rebasing and testing: https://github.com/mono/CppSharp/pull/983
If I recall, it was close to done, but never got the last push it needed to land. 

My bad, looks like you're right, `std::map` was never implemented for C++/CLI. They are not blittable, but since you're using C++/CLI, it should be relatively easy to implement marshaling based on the code from `std::vector` type map.

Feel free to come by our Gitter channel if you need further help.

**Comment by zillemarco** [comment]
@tritao I'm really sorry I wasn't able to finish that. I got stuck with other issues at work and I couldn't find any time to give it the last push. I hope to be able to come back to it since I'm going to need that at work sooner or later, so I'll use that chance to close that chapter.
